prusti-specs/src/specifications/common.rs

```diff
@@ -111,6 +111,34 @@ impl Into<u64> for ExpressionId {
     }
 }
 
+pub(crate) struct NameGenerator {}
+
+impl NameGenerator {
+    pub(crate) fn new() -> Self { Self { } }
+    pub(crate) fn generate_struct_name(&self, item: &syn::ItemImpl) -> Result<String, String> {
+        let mut path_str: String = String::new();
+
+        match &*item.self_ty {
+            syn::Type::Path (ty_path) => {
+                for seg in ty_path.path.segments.iter() {
+                    path_str.push_str(&seg.ident.to_string());
+                }
+            }
+            _ => {
+                return Err("expected a path".to_string());
+            }
+        };
+        let uuid = Uuid::new_v4().to_simple();
+
+        Ok(format!("PrustiStruct{}{}", path_str, uuid))
+    }
+
+    pub(crate) fn generate_mod_name(&self, ident: &syn::Ident) -> String {
+        let uuid = Uuid::new_v4().to_simple();
```

Comment:
`NameGenerator` could use an internal counter (e.g. with `RefCell`) instead of calling `Uuid::new_v4()` to generate fresh names. This is necessary (but not sufficient) to make Prusti's encoding deterministic.

Reply:
I *think* there are two reasons UUID is used:

 - `NameGenerator` is not shared, it is recreated per `mod` and `struct`. Without a global static state there may be "internal" collisions.
 - The external spec modules and structs are renamed and rewritten, but they still exist in the namespace. Renaming the module to `original_name<UUID>` is less likely to result in name collisions with actual code than `original_name1`, `original_name2`, etc.

Nevertheless, I don't see a reason why modules can't be renamed just like `structs`, with a `PrustiMod` prefix…

> This is necessary (but not sufficient) to make Prusti's encoding deterministic.

Could you elaborate on that? Did you have other things in mind to make it "sufficient"?

Reply:
> Could you elaborate on that? Did you have other things in mind to make it "sufficient"?

I mean that Prusti is already not deterministic for reasons that are orthogonal this PR. For example, the Viper code that we currently generate often contains `closure$...` predicates that have a seemingly random integer in their name.

Reply:
I found that the UUID technique is also used by `prusti-specs` work, and I don't have a nice solution to the problem of sharing a counter between procedural macros. So, I'm fine to leave this as it is.